### PR TITLE
[Slurm] Add CustomSlurmSettings parameter for queue and compute resource

### PR DESCRIPTION
**Pull request by NSsirena**
### Description of changes
* Add optional parameter `CustomSlurmSettings` for queue and compute resource
* Add update policy for `CustomSlurmSettings` parameter
* Add validator for CustomSlurmSettings
* Add Unit Tests for CustomSlurmSettings

### Tests
* local dryrun with config 
```
  SlurmQueues:
  - Name: queue1
    CustomSlurmSettings:
      Nodes: Node1
      PartitionName: CustomPartition
    ComputeResources:
    - Name: t2micro
      CustomSlurmSettings:
        NodeName: CustomNodeName
        State: CustomState
```

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] ~~Check if documentation is impacted by this change.~~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/validators/slurm_settings_validator.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (jdeamicis) [inline cli/src/pcluster/validators/slurm_settings_validator.py]: I wonder if we have to use 2023 now?

File: cli/src/pcluster/config/cluster_config.py [hunk 2]
```diff
@@ -191,6 +191,13 @@
     SupportedVersionsValidator,
     UserNameValidator,
 )
+from pcluster.validators.slurm_settings_validator import (
+    SLURM_SETTINGS_COMPUTE_RESOURCE_DENY_LIST,
+    SLURM_SETTINGS_QUEUE_DENY_LIST,
```
Comment (jdeamicis) [inline cli/src/pcluster/config/cluster_config.py]: Should we use `CUSTOM_SLURM_SETTINGS_*` instead of `SLURM_SETTINGS_*`?

File: cli/src/pcluster/config/cluster_config.py [hunk 3]
```diff
@@ -1935,7 +1944,11 @@ def __init__(self, instance_type: str, **kwargs):
 class SlurmFlexibleComputeResource(_BaseSlurmComputeResource):
     """Represents a Slurm Compute Resource with Multiple Instance Types."""
 
-    def __init__(self, instances: List[FlexibleInstanceType], **kwargs):
+    def __init__(
+        self,
+        instances: List[FlexibleInstanceType],
+        **kwargs,
+    ):
```
Comment (jdeamicis) [inline cli/src/pcluster/config/cluster_config.py:1950]: Nitpick: should we revert this change, since it doesn't do anything?

File: cli/src/pcluster/config/cluster_config.py [hunk 4]
```diff
@@ -1972,7 +1985,11 @@ def max_network_interface_count(self) -> int:
 class SlurmComputeResource(_BaseSlurmComputeResource):
     """Represents a Slurm Compute Resource with a Single Instance Type."""
 
-    def __init__(self, instance_type, **kwargs):
+    def __init__(
+        self,
+        instance_type,
+        **kwargs,
+    ):
```
Comment (jdeamicis) [inline cli/src/pcluster/config/cluster_config.py:1991]: Nitpick: should we revert this change, since it doesn't do anything?

File: cli/src/pcluster/config/cluster_config.py [hunk 5]
```diff
@@ -2161,9 +2178,11 @@ class SlurmQueue(_CommonQueue):
     def __init__(
         self,
         allocation_strategy: str = None,
+        custom_slurm_settings: Dict = None,
         **kwargs,
     ):
         super().__init__(**kwargs)
+        self.custom_slurm_settings = Resource.init_param(custom_slurm_settings, default={})
```
Comment (jdeamicis) [inline cli/src/pcluster/config/cluster_config.py:2184]: I'm wondering when we need `Resource.init_param()` and when not...

File: cli/src/pcluster/config/cluster_config.py [hunk 6]
```diff
@@ -2214,6 +2233,14 @@ def _register_validators(self, context: ValidatorContext = None):
                 queue_name=self.name,
                 subnet_ids=self.networking.subnet_ids,
             )
+        if bool(self.custom_slurm_settings):  # if not empty register validator
```
Comment (jdeamicis) [inline cli/src/pcluster/config/cluster_config.py]: Nitpick: `bool()` should not be necessary here.

File: cli/src/pcluster/config/cluster_config.py [hunk 7]
```diff
@@ -2231,6 +2258,14 @@ def _register_validators(self, context: ValidatorContext = None):
                 is False,
                 multi_az_enabled=self.multi_az_enabled,
             )
+            if bool(compute_resource.custom_slurm_settings):  # if not empty register validator
```
Comment (jdeamicis) [inline cli/src/pcluster/config/cluster_config.py]: Same as above.

File: cli/src/pcluster/validators/slurm_settings_validator.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+from enum import Enum
+
+from pcluster.validators.common import FailureLevel, Validator
+
+SLURM_SETTINGS_QUEUE_DENY_LIST = ["Nodes", "State", "PartitionName", "ResumeTimeout", "SuspendTime"]
+SLURM_SETTINGS_COMPUTE_RESOURCE_DENY_LIST = ["NodeName", "NodeHostName", "NodeAddr", "State", "Gres"]
```
Comment (jdeamicis) [inline cli/src/pcluster/validators/slurm_settings_validator.py]: We have to double-check if the parameters in the `slurm.conf` are case-sensitive. If they aren't, we may have to do this validation case-insensitive.

Comment (jdeamicis) [reply cli/src/pcluster/validators/slurm_settings_validator.py]: Second paragraph from the [`slurm.conf` documentation](https://slurm.schedmd.com/slurm.conf.html#SECTION_DESCRIPTION):

> The contents of the file are case insensitive except for the names of nodes and partitions.

File: cli/src/pcluster/validators/slurm_settings_validator.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class SlurmCustomSettingLevel(str, Enum):
```
Comment (jdeamicis) [inline cli/src/pcluster/validators/slurm_settings_validator.py:23]: What is the `str` here used for?

Comment (jdeamicis) [reply cli/src/pcluster/validators/slurm_settings_validator.py:23]: So the object in the class inherit from both Enums and strings?

File: cli/src/pcluster/validators/slurm_settings_validator.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    def _validate(self, custom_settings, deny_list, settings_level: SlurmCustomSettingLevel):
+        denied_settings = set()
+
+        for setting in deny_list:
+            if setting in custom_settings.keys():
+                denied_settings.add(setting)
+        if len(denied_settings) > 0:
+            settings = ",".join(sorted(denied_settings))
+            self._add_failure(
+                f"Setting or overriding the following custom Slurm settings at {settings_level} level is not allowed: "
```
Comment (jdeamicis) [inline cli/src/pcluster/validators/slurm_settings_validator.py]: I would word it "Using the following custom Slurm settings...".

File: cli/src/pcluster/validators/slurm_settings_validator.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+    QUEUE = "Queue"
+    COMPUTE_RESOURCE = "ComputeResource"
+
+
+class SlurmCustomSettingsValidator(Validator):
+    """
+    Slurm Custom Settings validator.
+
+    Validate custom settings in Slurm ComputeResource and queue.
```
Comment (jdeamicis) [inline cli/src/pcluster/validators/slurm_settings_validator.py]: Nitpick: "Queue" instead of "queue".

File: cli/tests/pcluster/validators/test_cluster_validators.py [hunk 12]
```diff
@@ -187,6 +188,45 @@ def test_cluster_name_validator_slurm_accounting(cluster_name, scheduling, shoul
...
+            SlurmCustomSettingLevel.COMPUTE_RESOURCE,
+            "",
+        ),
+        (
+            "Fails when custom settings are in the deny_list, invalid parameters are reported",
+            {"Denied1": "Value1", "Denied2": "Value2"},
+            ["Denied1", "Denied2"],
+            SlurmCustomSettingLevel.COMPUTE_RESOURCE,
+            "Setting or overriding the following custom Slurm settings at"
+            " ComputeResource level is not allowed: Denied1,Denied2",
```
Comment (jdeamicis) [inline cli/tests/pcluster/validators/test_cluster_validators.py]: For better clarity, please put the full sentence here.

File: cli/tests/pcluster/validators/test_cluster_validators.py [hunk 13]
```diff
@@ -187,6 +188,45 @@ def test_cluster_name_validator_slurm_accounting(cluster_name, scheduling, shoul
     assert_failure_messages(actual_failures, expected_message)
 
 
+@pytest.mark.parametrize(
```
Comment (jdeamicis) [inline cli/tests/pcluster/validators/test_cluster_validators.py:195]: I'm not seeing a unit test on the generic warning. Could we add one?

**Review by jdeamicis — CHANGES_REQUESTED** [review]
Good job. We need to sort out the case-insensitivity of the Slurm configuration files.

I left some minor comments around the code.

File: cli/src/pcluster/validators/slurm_settings_validator.py (hunk 9 above)
Comment (NSsirena) [reply cli/src/pcluster/validators/slurm_settings_validator.py:23]: Yes. 
In python 3.11 there is an StrEnum and you can use that.

Here to use a string as enum value we have to inherit from both.

File: cli/src/pcluster/config/cluster_config.py (hunk 2 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py]: They are `SLURM_SETTINGS` meaning that they are defined by Slurm and we are `denying` their use.

Using here `CUSTOM` gives the impression that they are defined by the user, like a tag or something.

File: cli/src/pcluster/config/cluster_config.py (hunk 3 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py:1950]: I thought about it, but then I decided to keep them.
Having the same `init style` across all resources make the code more readable.

File: cli/src/pcluster/config/cluster_config.py (hunk 4 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py:1991]: same as above.

File: cli/src/pcluster/config/cluster_config.py (hunk 5 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py:2184]: Not 100% sure, my understanding is that we use it when we want to have a default value and avoid `None`

File: cli/src/pcluster/config/cluster_config.py (hunk 6 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py]: It is.
I had to add it ..without it doesn't work with empty dictionaries.

File: cli/src/pcluster/config/cluster_config.py (hunk 7 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py]: same

File: cli/src/pcluster/validators/slurm_settings_validator.py (hunk 8 above)
Comment (NSsirena) [reply cli/src/pcluster/validators/slurm_settings_validator.py]: Good catch!
I'll update the check!

File: cli/src/pcluster/validators/slurm_settings_validator.py (hunk 10 above)
Comment (NSsirena) [reply cli/src/pcluster/validators/slurm_settings_validator.py]: I picked the working from the design doc, but I can update it if you think this version is better.
In this case I'd update also the doc.

File: cli/tests/pcluster/validators/test_cluster_validators.py (hunk 12 above)
Comment (NSsirena) [reply cli/tests/pcluster/validators/test_cluster_validators.py]: Not sure I got your point.
It is the full sentence, I just had to split it in two lines to make the code-linter happy

File: cli/tests/pcluster/validators/test_cluster_validators.py (hunk 13 above)
Comment (NSsirena) [reply cli/tests/pcluster/validators/test_cluster_validators.py:195]: It is a different validator ..not sure what to test about it apart from its presence/absence.
I'll try to figure out a way to test it.

File: cli/src/pcluster/config/cluster_config.py (hunk 2 above)
Comment (jdeamicis) [reply cli/src/pcluster/config/cluster_config.py]: I'm fine with that.

Just had an idea: can we have one single `SLURM_SETTINGS_DENY_LIST` dictionary so that we can easily extend the structure in the future?

So we will have something like:

```
SLURM_SETTINGS_DENY_LIST = {
    "Queue": [ ... ],
    "ComputeResource": [ ... ],
    "SlurmConf": [ ... ],
    ...
}
```

Please feel free to propose improvements.

File: cli/src/pcluster/validators/slurm_settings_validator.py (hunk 8 above)
Comment (NSsirena) [reply cli/src/pcluster/validators/slurm_settings_validator.py]: Fixed.
I also added a specific unit test.

File: cli/src/pcluster/validators/slurm_settings_validator.py (hunk 10 above)
Comment (NSsirena) [reply cli/src/pcluster/validators/slurm_settings_validator.py]: Fixed

File: cli/src/pcluster/validators/slurm_settings_validator.py (hunk 11 above)
Comment (NSsirena) [reply cli/src/pcluster/validators/slurm_settings_validator.py]: Fixed

File: cli/tests/pcluster/validators/test_cluster_validators.py (hunk 13 above)
Comment (NSsirena) [reply cli/tests/pcluster/validators/test_cluster_validators.py:195]: I added a test for the Warning

File: cli/tests/pcluster/validators/test_cluster_validators.py (hunk 12 above)
Comment (NSsirena) [reply cli/tests/pcluster/validators/test_cluster_validators.py]: With the rewording fits into one line, so it should be easier to check that it is the full sentence.

File: cli/src/pcluster/validators/slurm_settings_validator.py (hunk 1 above)
Comment (NSsirena) [reply cli/src/pcluster/validators/slurm_settings_validator.py]: Changed to 2023

File: cli/src/pcluster/config/cluster_config.py (hunk 2 above)
Comment (NSsirena) [reply cli/src/pcluster/config/cluster_config.py]: Done. 
I've also added new values in the deny lists.

**Review by jdeamicis — DISMISSED** [review]

File: cli/src/pcluster/config/cluster_config.py (hunk 6 above)
Comment (jdeamicis) [reply cli/src/pcluster/config/cluster_config.py]: That's weird. I'm not getting this behavior locally:
```
In [1]: A = {}

In [2]: if A:
   ...:     print("yes")
   ...: 

In [3]: B = {"bla": "blu"}

In [4]: if B:
   ...:     print("yes")
   ...: 
yes

```

**Review by jdeamicis — APPROVED** [review]
